Importing necessary libraries

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix, accuracy_score

Reading the dataset with Pandas after mounting the dataset in google colab

In [ ]:
#reading the dataset with a dataframe
df = pd.read_csv("final_preprocessed_dataset.csv")
df=df.drop('Unnamed: 0',1)

In [ ]:
#eye balling the data
df.head(5)

Resource  ... APTGroup
0  4d74c8da7274bb56bab01f47334a9551f57a6dcf  ...        9
1  cc221465dac981f4934fef39d41ddb2e1d26299f  ...        3
2  a49718feddf874a62049809af51e858ad41897c4  ...        0
3  78c00614535b949770708069615764d0155cad76  ...       10
4  44787ddf91b10291f8338590c5d99839040e1cd5  ...        3

[5 rows x 151 columns]

Dropping resource column and categorical APT group column

In [ ]:
df = df.drop('Resource', 1)
df =df.drop('APT Group',1)

In [ ]:
#eye balling information about our dataset
df.head(5)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2862 entries, 0 to 2861
Columns: 149 entries, Entry Point to APTGroup
dtypes: float64(1), int64(148)
memory usage: 3.3 MB


Sklearn Train Test Split Method : Splitting the data into training and testing data.

In [ ]:
#the labelled column is Apt Group and the rest is features
label = np.array(df['APTGroup'])
features = df.drop('APTGroup',1)

In [ ]:
features.head()


Entry Point  'ACTIVEDS.DLL'  ...  'SWEDISH'  'TURKISH DEFAULT'
0      54299.0               0  ...          0                  0
1       7627.0               0  ...          0                  0
2      38026.0               0  ...          0                  0
3     396884.0               0  ...          0                  0
4       6948.0               0  ...          0                  0

[5 rows x 148 columns]

In [ ]:
label

array([ 9,  3,  0, ..., 10,  5, 10])

In [ ]:
#choosing a 70-30 split to test out the performance
from sklearn.model_selection import train_test_split
seed =50
X_train, X_test, y_train, y_test = train_test_split(features,label,test_size=0.30, random_state = seed)

In [ ]:
X_train.head(5)

Entry Point  'ACTIVEDS.DLL'  ...  'SWEDISH'  'TURKISH DEFAULT'
800       83741.0               0  ...          0                  0
2684    1079966.0               0  ...          0                  0
64        24146.0               0  ...          0                  0
1010     180864.0               0  ...          0                  0
2360     430686.0               0  ...          0                  0

[5 rows x 148 columns]

In [ ]:
X_test.head(5)

Entry Point  'ACTIVEDS.DLL'  ...  'SWEDISH'  'TURKISH DEFAULT'
1688      23536.0               0  ...          0                  0
2455      10136.0               0  ...          0                  0
2716     470572.0               0  ...          0                  0
2296      10710.0               0  ...          0                  0
578      549060.0               0  ...          0                  0

[5 rows x 148 columns]

Building the Random Forest Classifier

In [ ]:
rf_classifier = RandomForestClassifier(
                      min_samples_leaf=50,
                      n_estimators=150,
                      bootstrap=True,
                      oob_score=True,
                      n_jobs=-1,
                      random_state=seed,
                      max_features='auto')

In [ ]:
rf_classifier.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=50, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=150,
                       n_jobs=-1, oob_score=True, random_state=50, verbose=0,
                       warm_start=False)

In [ ]:
y_pred = rf_classifier.predict(X_test)

In [ ]:
print(y_pred)

[ 0 11 10 11 10  9 10  9  0  7  9 10  0 11 11  0  9  9  9  0  9  9 10  7
  7 11 11 10 10  0  0 11 11  9  0  8 10  0 11  7 11 10  0  0 11  0 11  0
  9 11 11  0  5  9 11  0 11 10 11 11  7  1  0  0 10  0 11 11 11  7 10  8
  0  9 11  9 10 11 11 10 10  0 11 11 11  0  9  0 11  8  7  9 11  0  9  7
 11 10 11  9  9  5 10  8 11 11 11  7 11  8 11 11  8 11  9  9  8  9  7 10
  7  1 11  0 10 11  7  5  0  0  0  9 11 10 11  0  7 10 11  0  0  0  0  7
 11 11 11  0  9 11  0 11 10 11  8 10  0  0 11 11  0  9 10 11 10  9 11  9
 10 11 11 11  7 11  9 11  9 11 11  0 10 11  0 10 11  0  9  0 11 10 11  0
  9 11 10  1  8 11  0 10  9  9 10 11  9  0 10 10  9  0 10  0 11 10 11 10
  0 10 10  9 11 10  9  0  5  0  0  0 11  0 10 11  7  9  9  0 11 11  1 11
 10  7  7 11  0 11 10 10  7 10  0  0 11  7 11 11 11 11  9  0 10 10  0 10
  0 11  8 11  0  8 10 11  0  0 11 10 10 11 11  7 11  7  7 10  0  9  0  9
  0 11 11  9  0  7 10 11  0  8  0  8  0  9  7 11  0  9  0 11  7  9 10 11
  9 10 11  0 11  9 11  0 11  9 10  0  0 11  0  0 11

Confusion Matrix

In [ ]:
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[107   0   0   0   0   0   0   2   2   5   0  10]
 [  2  17   0   0   0   0   0   2   0   4   2  35]
 [  4   0   0   0   0   0   0   4   0   0   1   1]
 [  0   0   0   0   0   0   0   0   0   0   1  27]
 [ 18   0   0   0   0   0   0   1   0   1   2  24]
 [ 13   0   0   0   0  10   0   5   3   1  13  36]
 [ 29   0   0   0   0   0   0   1   0   1   0  19]
 [  9   0   0   0   0   0   0  43   0   2  17  15]
 [  1   0   0   0   0   0   0   0  39   0   1   0]
 [  0   0   0   0   0   0   0   0   0 113   0   0]
 [  2   0   0   0   0   0   0   0   0   2  90   0]
 [ 36   0   0   0   0   0   0   0   0   0   6  80]]


0.580908032596042

# Observation:

Accuracy acquired about 58% accuracy. Needs more hyperparameter tuning to increase the accuracy. Maybe contributing other factors along the way too can lead to better accuracy.

## Model Attempt#2 : Hyper Parameter tunning

In [ ]:
"""
rf_classifier = RandomForestClassifier(
                      min_samples_leaf=50,
                      n_estimators=150,
                      bootstrap=True,
                      oob_score=True,
                      n_jobs=-1,
                      random_state=seed,
                      max_features='auto')
                      """
rf_classifier = RandomForestClassifier(
                      min_samples_leaf=3,
                      n_estimators=300,
                      bootstrap=True,
                      oob_score=True,
                      n_jobs=-1,
                      random_state=seed,
                      max_features='auto')

In [ ]:
rf_classifier.fit(X_train, y_train)
y_pred = rf_classifier.predict(X_test)
print("accuracy ", accuracy_score(y_test, y_pred) )

accuracy  0.830034924330617


#Cross validation 

In [ ]:
df.head(5)

Entry Point  'ACTIVEDS.DLL'  ...  'TURKISH DEFAULT'  APTGroup
0      54299.0               0  ...                  0         9
1       7627.0               0  ...                  0         3
2      38026.0               0  ...                  0         0
3     396884.0               0  ...                  0        10
4       6948.0               0  ...                  0         3

[5 rows x 149 columns]

In [ ]:
 from sklearn.model_selection import ShuffleSplit
 ss = ShuffleSplit(n_splits=20, test_size=0.30)
accuracy=0
for train_indices, test_indices in ss.split(df):
  #print("Train")
  #print(X_train)
  #print("Test")
  #print(X_test)
  
  train = df.iloc[train_indices]
  y_train= np.array(train['APTGroup'])
  X_train= train.drop('APTGroup',1)
  

  test = df.iloc[test_indices]
  y_test= np.array(train['APTGroup'])
  X_test= train.drop('APTGroup',1)
  
  rf_classifier.fit(X_train, y_train)
  y_pred = rf_classifier.predict(X_test)
  print("accuracy ", accuracy_score(y_test, y_pred) )
  accuracy+=accuracy_score(y_test, y_pred)

print(accuracy/20)


accuracy  0.8582126809785322
accuracy  0.8666999500748876
accuracy  0.872191712431353
accuracy  0.8691962056914628
accuracy  0.8592111832251622
accuracy  0.8681977034448327
accuracy  0.8572141787319022
accuracy  0.8696954568147778
accuracy  0.872191712431353
accuracy  0.8642036944583126
accuracy  0.8587119321018473
accuracy  0.8666999500748876
accuracy  0.871193210184723
accuracy  0.8706939590614079
accuracy  0.8612081877184223
accuracy  0.8731902146779831
accuracy  0.8756864702945582
accuracy  0.8607089365951074
accuracy  0.8666999500748876
accuracy  0.8627059410883674
0.8662256615077384


# Observation 1:

Accuracy acquired : 86.62 % for n_splits=20 \\

# Another method for cross validation: cross_val_score


In [ ]:
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import RandomForestClassifier
print(cross_val_score(rf_classifier, features, label, cv=20, scoring ='accuracy').mean())

0.8420600233100233


# Observation 2:

Accuracy acquired : 84.2 % for n_splits=20 \\

## Important Feature Calculation

In [ ]:
rf_classifier.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=3, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=300,
                       n_jobs=-1, oob_score=True, random_state=50, verbose=0,
                       warm_start=False)

In [ ]:
df.head(3)

Entry Point  'ACTIVEDS.DLL'  ...  'TURKISH DEFAULT'  APTGroup
0      54299.0               0  ...                  0         9
1       7627.0               0  ...                  0         3
2      38026.0               0  ...                  0         0

[3 rows x 149 columns]

In [ ]:
pd.set_option("display.max.rows", None)

In [ ]:
# Important Features
feature_importance = pd.DataFrame(rf_classifier.feature_importances_,
                                   index = X_train.columns,
                                    columns=['importance']).sort_values('importance',ascending=False)
feature_importance
# summarize feature importance
#for i,v in enumerate(importance):
 #   print('Feature: %0d, Score: %.5f' % (i,v))

importance
Entry Point                            0.119079
'NEUTRAL'                              0.104249
'MSVCRT.DLL'                           0.069596
'WS2_32.DLL'                           0.058632
'WININET.DLL'                          0.057716
'USER32.DLL'                           0.041985
'ENGLISH US'                           0.033963
'OLE32.DLL'                            0.033411
'SHLWAPI.DLL'                          0.032455
'SHELL32.DLL'                          0.031832
'CRYPT32.DLL'                          0.030989
'CHINESE SIMPLIFIED'                   0.029964
'KERNEL32.DLL'                         0.026748
'VERSION.DLL'                          0.024052
'ADVAPI32.DLL'                         0.023814
'COMCTL32.DLL'                         0.023397
'MSCOREE.DLL'                          0.022960
'GDI32.DLL'                            0.022219
'OLEAUT32.DLL'                         0.021467
'MPR.DLL'                              0.020085
'IPHLPAPI.DLL'                         0.017153
'URLMON.DLL'                           0.015682
'WSOCK32.DLL'                          0.014410
'COMDLG32.DLL'                         0.010005
'MFC42.DLL'                            0.009786
'SECUR32.DLL'                          0.008959
'KOREAN'                               0.007737
'PSAPI.DLL'                            0.007596
'MSVBVM60.DLL'                         0.007147
'MSVCP60.DLL'                          0.006978
'NETAPI32.DLL'                         0.006416
'WTSAPI32.DLL'                         0.004943
'USERENV.DLL'                          0.004611
'WINMM.DLL'                            0.004376
'IMAGEHLP.DLL'                         0.003934
'ENGLISH UK'                           0.003495
'DNSAPI.DLL'                           0.003479
'NEUTRAL DEFAULT'                      0.003387
'NTOSKRNL.EXE'                         0.003107
'WINSPOOL.DRV'                         0.002738
'RPCRT4.DLL'                           0.002672
'GERMAN'                               0.002500
'GDIPLUS.DLL'                          0.002467
'NETMGR.DLL'                           0.002140
'RUSSIAN'                              0.002082
'LZ32.DLL'                             0.001344
'WINHTTP.DLL'                          0.001257
'JAPANESE DEFAULT'                     0.001104
'HAL.DLL'                              0.001084
'OLEDLG.DLL'                           0.001079
'MSVCR80.DLL'                          0.000950
'TURKISH DEFAULT'                      0.000819
'AVICAP32.DLL'                         0.000773
'IMM32.DLL'                            0.000683
'NTDLL.DLL'                            0.000654
'OLEACC.DLL'                           0.000549
'OPENGL32.DLL'                         0.000546
'GLU32.DLL'                            0.000528
'MSIMG32.DLL'                          0.000523
'SETUPAPI.DLL'                         0.000308
'FRENCH'                               0.000291
'MSVCR90.DLL'                          0.000246
'CHINESE TRADITIONAL'                  0.000240
'FLTMGR.SYS'                           0.000198
'OLEPRO32.DLL'                         0.000172
'UXTHEME.DLL'                          0.000106
'DSOUND.DLL'                           0.000098
'D3D9.DLL'                             0.000017
'DDRAW.DLL'                            0.000016
'WLDAP32.DLL'                          0.000000
'D3DX9_32.DLL'                         0.000000
'WINTRUST.DLL'                         0.000000
'DIVEHI NEUTRAL'                       0.000000
'CZECH DEFAULT'                        0.000000
'WWUTILS.DLL'                          0.000000
'CRT.DLL'                              0.000000
'XUL.DLL'                              0.000000
'AVIFIL32.DLL'                         0.000000
'COMSVCS.DLL'                          0.000000
'AZERI LATIN'                          0.000000
'CABINET.DLL'                          0.000000
'ALBANIAN DEFAULT'                     0.000000
'ENGLISH AUS'                          0.000000
'D3DX

Exporting the results into an excel file for better view

In [ ]:
excel_data = feature_importance.copy()

In [ ]:
excel_data.to_csv('important_features.csv', index=True)

In [ ]:
%ls

final_preprocessed_data.csv     important_features.csv
final_preprocessed_dataset.csv  sample_data/


In [ ]:
#checking with shell command if there was a proper write in the new file created
! cat important_features.csv

,importance
Entry Point,0.11907935485280088
'NEUTRAL',0.10424881306877244
'MSVCRT.DLL',0.0695956171704782
'WS2_32.DLL',0.05863191298620216
'WININET.DLL',0.05771601751581001
'USER32.DLL',0.041985256891548535
'ENGLISH US',0.03396263127410925
'OLE32.DLL',0.03341085388898082
'SHLWAPI.DLL',0.03245535002135421
'SHELL32.DLL',0.03183228205120376
'CRYPT32.DLL',0.03098899174654575
'CHINESE SIMPLIFIED',0.029964193826814443
'KERNEL32.DLL',0.02674821228034456
'VERSION.DLL',0.02405199710555099
'ADVAPI32.DLL',0.023814258924715677
'COMCTL32.DLL',0.02339712023294005
'MSCOREE.DLL',0.022960118558232576
'GDI32.DLL',0.022219215029960618
'OLEAUT32.DLL',0.021466627134163307
'MPR.DLL',0.020084569411950172
'IPHLPAPI.DLL',0.017153112223686314
'URLMON.DLL',0.015682292500798646
'WSOCK32.DLL',0.014409861845197816
'COMDLG32.DLL',0.010004560966906818
'MFC42.DLL',0.009786070476067828
'SECUR32.DLL',0.008958750081738878
'KOREAN',0.007736578064476168
'PSAPI.DLL',0.007596414834916762
'MSVBVM60.DLL',0.007147067859624855
'